### PR TITLE
`Make up/down` for services

### PR DESCRIPTION
**Pull request by EliChin**
- `make up.basic` builds up basic neofs
- `make up.service` needs `val=s3_gate\http_gate\coredns` par to build up certain service
- `make down.service` needs `val=s3_gate\http_gate\coredns` par to stop certain service
- in `services/s3_gate/docker-compose.yml` name of the service changed to the relevant one

#154 
Signed-off-by: Elizaveta Chichindaeva <elizaveta@nspcc.ru>

### CODE REVIEW

**Comment by realloc** [comment]
I'd' suggest using a different approach, similar to adding docker wrapper in [neofs-node Makefile](https://github.com/nspcc-dev/neofs-node/blob/master/Makefile#L97), so that we could:

- `make up/basic` - to start just basic NeoFS services
- `make up/<service>` - to start a specific service.

The same with `make down`. The default `make all` can be left as it is now.

File: Makefile [hunk 1]
```diff
@@ -57,12 +61,32 @@ up: pull get vendor/hosts
...
+#Build up NeoFS
+.PHONY: up.basic
+up.basic: pull get vendor/hosts
+	@$(foreach SVC, $(START_NEOFS), $(shell docker-compose -f services/$(SVC)/docker-compose.yml up -d))
+	@./bin/tick.sh
+	@echo "Basic NeoFS Developer Environment is ready"
+
+# Build up certain service (make up.service val=service)
+.PHONY: up.service
+up.service: pull get vendor/hosts $(START_BASENET)
```
[the remark this answers is not in the document]
Comment (aprasolova) [reply Makefile:74]: I suppose, `make up` and `make down` do it, aren't they?

[the remark this answers is not in the document]
Comment (carpawell) [reply Makefile:74]: I mean exactly "services": `http`, `s3`

I do not know the real reasons for that PR creation so I may be wrong somewhere: I see it like I should be able to make two things: 1. `make base` and I have core working, 2. I want to add additional "services" to core and I do `make up.services`

**Comment by EliChin** [comment]
@realloc  changed according to your suggestion

File: Makefile [hunk 2]
```diff
@@ -57,12 +61,32 @@ up: pull get vendor/hosts
 	@./bin/tick.sh
 	@./bin/config.sh string SystemDNS container
 	@echo "NeoFS Developer Environment is ready"
+	
+#Build up NeoFS
```
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: Done

File: Makefile [hunk 3]
```diff
@@ -57,12 +61,32 @@ up: pull get vendor/hosts
...
+.PHONY: up.basic
+up.basic: pull get vendor/hosts
```
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: Done.

Get is only made for basic neofs services anyway, not for s3 and http services. 
`pull` can be now used with certain services as `pull/<service>`

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: 1. You can have core working with `make up/basic`. 2. If you need to start both http and s3 at once, it's quite easy to do `make up/s3_gate up/http_gate`.

[the remark this answers is not in the document]
Comment (carpawell) [reply Makefile]: 2. yes, I got it. it is clear and ok _if it was planned_ to up "services" one by one

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply Makefile]: > Get is only made for basic neofs services anyway

yes, but I don't think it is the rule

**Review by realloc — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: It was :)

File: Makefile [hunk 4]
```diff
@@ -45,6 +46,12 @@ pull:
 	$(foreach SVC, $(PULL_SVCS), $(shell cd services/$(SVC) && docker-compose pull))
 	@:
 
+# Pull certain Docker image
+.PHONY: pull/%
```
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile:50]: Do you suggest to make "full" `pull` while making `up/<service>`?

File: Makefile [hunk 5]
```diff
@@ -58,11 +65,36 @@ up: pull get vendor/hosts
 	@./bin/config.sh string SystemDNS container
 	@echo "NeoFS Developer Environment is ready"
 
+# Start basenet
+.PHONY: basenet/%
```
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: It's made separate to allow building up a service independently.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (realloc) [reply Makefile:50]: Yes. That's done just once, NP with that.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (realloc) [reply Makefile]: You can't run any of the services without basic NeoFS and basenet is the requirement for it.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: Done but not relevant with this [PR](https://github.com/nspcc-dev/neofs-dev-env/pull/157)

File: Makefile [hunk 6]
```diff
@@ -27,6 +27,7 @@ PULL_SVCS = $(shell find ./services -type f -name 'docker-compose.yml' | sort -u
 
 # List of services to run
 START_SVCS = $(shell cat .services | grep -v \\\#)
+START_NEOFS = $(shell cat .services | grep -v -e .*gate -e \\\# -e coredns)
```
[the remark this answers is not in the document]
Comment (EliChin) [reply Makefile]: Done

File: Makefile [hunk 7]
```diff
@@ -27,6 +27,7 @@ PULL_SVCS = $(shell find ./services -type f -name 'docker-compose.yml' | sort -u
 
 # List of services to run
 START_SVCS = $(shell cat .services | grep -v \\\#)
+START_NEOFS = $(shell cat .services | grep -e .*chain -e ir -e storage -e basenet)
```
Comment (realloc) [inline Makefile:30]: What is we create a list of basic services, instead of doing exclusion?

Comment (EliChin) [reply Makefile:30]: It's not exclusion. We `grep` morph_chain, chain, storage, ir, and basenet4 thus, we receive a list of basic services.

Comment (realloc) [reply Makefile:30]: The initial idea behind service directories is that they are independent and can be added by developer. If, as a developer, I add a new service, I also must edit the Makefile, and that's a bad thing to do. That's counterintuitive and leads to occasionally broken devenv.

File: bin/tick.sh [hunk 8]
```diff
@@ -5,7 +5,7 @@
 . services/ir/.ir.env
 
 # NeoGo binary path.
-NEOGO="${NEOGO:-docker exec -it main_chain neo-go}"
```
Comment (aprasolova) [inline bin/tick.sh:8]: Roll it back please and set `NEOGO` in the environment where you set up devenv instead

File: bin/tick.sh [hunk 9]
```diff
@@ -32,6 +32,7 @@ http://morph_chain.${LOCAL_DOMAIN}:30333 \
 ${NETMAP_ADDR} \
 epoch | grep 'value' | awk -F'"' '{ print $4 }'`
 
+echo ${NETMAP_ADDR}, ${ADDR}
```
Comment (aprasolova) [inline bin/tick.sh:35]: Remove this please

File: Makefile [hunk 10]
```diff
@@ -58,11 +59,30 @@ up: get vendor/hosts
...
+up/basic: get vendor/hosts
+	@$(foreach SVC, $(START_NEOFS), $(shell docker-compose -f services/$(SVC)/docker-compose.yml up -d))
+	@./bin/tick.sh
+	@echo "Basic NeoFS Developer Environment is ready"
+
+# Build up certain service
+.PHONY: up/%
+up/%: get vendor/hosts
+	@docker-compose -f services/$*/docker-compose.yml up -d
+	@./bin/tick.sh
```
Comment (alexvanin) [inline Makefile:73]: I guess if you want to start certain service, you don't have to tick epoch. It is not possible for services before `morph_chain`, e.g. try `make up/basenet`.

**Review by alexvanin — APPROVED** [review]

**Review by anatoly-bogatyrev — APPROVED** [review]

**Review by realloc — CHANGES_REQUESTED** [review]
Wait please

**Review by realloc — APPROVED** [review]
Just to remove `request changes` status. Please review.

File: Makefile [hunk 11]
```diff
@@ -52,17 +54,43 @@ get: $(foreach SVC, $(GET_SVCS), get.$(SVC))
...
+down: down/add down/basic
+	@echo "Full NeoFS Developer Environment is down"
+
+.PHONY: down/add
+down/add:
 	$(foreach SVC, $(STOP_SVCS), $(shell docker-compose -f services/$(SVC)/docker-compose.yml down))
 
+# Stop basic environment
+.PHONY: down/basic
+down/basic:
```
Comment (alexvanin) [inline Makefile:86]: @realloc Don't see any use case in `down/basic` target. Actually, I prefer previous solution from @EliChin with two simple target: shutdown everything or shutdown specific service.

Comment (realloc) [reply Makefile:86]: The purpose of this target is to keep the previous `make down` behaviour without breaking the service stop order.

File: Makefile (hunk 10 above)
Comment (EliChin) [reply Makefile]: We suggest that before starting certain service with `make up/<service>`, basic services, including `morph_chain`, have been started already.

**Review by aprasolova — APPROVED** [review]

**Review by anatoly-bogatyrev — APPROVED** [review]

**Pull request merged**